sockeye/data_io.py

```diff
@@ -361,11 +372,14 @@ class ParallelBucketSentenceIter(mx.io.DataIter):
     :param buckets: List of buckets.
     :param batch_size: Batch_size of generated data batches.
            Incomplete batches are discarded if fill_up == None, or filled up according to the fill_up strategy.
+    :param batch_by_words: Size batches by words rather than sentences.
+    :param batch_num_devices: Number of devices batches will be parallelized across.
```

Comment:
I don't think I yet understand why we need this. 
Currently, the (sentence-based) batch size we get from the commandline is the 'total' batch size which then gets divided by n if we use n devices for training. Is this different for the word-based batch size? It'd be great to a have a minimum example to better understand this.

Reply:
We can add a comment about this to the code.

If we're only using whole sentences, batches will almost never have the right number of words.  If our batch size is 4096, we can add sentences until either we exceed 4096 or the next sentence would exceed 4096.  If all of our sentences are length 100, we could choose a batch size of 4000 or 4100.

When we have multiple devices, we also want the number of sentences to be evenly divisible so we won't waste computation (some devices processing N examples per batch, some N+1).

The two most direct solutions are:
1. Set number of sentences to the multiple of number of devices that gets us closest to the target batch size.  (Num devices can influence batch size)
2. Set number of sentences to be as close as possible to the target batch size and add padding sentences to make the batch size divisible by number of devices.  (Wasted computation for majority of batches)

This implementation goes with option 1.

Reply:
Thanks for the explanation! I agree that option 1 is the better choice.